### PR TITLE
ENT-8611 cleanup non-interactive stdin usage and dont use whitespace in passwo…

### PR DESCRIPTION
**Pull request by craigcomstock**
…rds, seems to foul up user account creation via API

### CODE REVIEW

**Comment by craigcomstock** [comment]
@cf-bottom jenkins builds please so I can test fr-tests again with new behavior, switch from getpass to input if --force-interactive is given (means do it even if no tty, like for automation via remote ssh).

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/8239/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/8239/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-8239/

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/8251/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/8251/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-8251/

**Comment by craigcomstock** [comment]
@cf-bottom jenkins for packages to test

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/8257/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/8257/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-8257/

**Comment by craigcomstock** [comment]
oops, had mentioned nova PR which is already merged :) another build without that: [Build Status](https://ci.cfengine.com/job/pr-pipeline/8258/)

**Comment by craigcomstock** [comment]
packages are ready, so retrying fr-tests with job 8258 packages: [Build Status](https://ci.cfengine.com/job/fr-tests/4652/)

**Comment by craigcomstock** [comment]
@cf-bottom jenkins

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/8262/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/8262/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-8262/

**Comment by craigcomstock** [comment]
buildslave in CI was broken, try again: [Build Status](https://ci.cfengine.com/job/pr-pipeline/8265/)

**Comment by craigcomstock** [comment]
retrying fr-tests with this build and the related deployment test changes: [Build Status](https://ci.cfengine.com/job/fr-tests/4656/)

**Comment by craigcomstock** [comment]
@cf-bottom no tests pipeline if you know what I mean :)

**Comment by craigcomstock** [comment]
[Build Status](https://ci.cfengine.com/job/pr-pipeline/8270/) to get packages, then will re-test with fr-tests.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/8271/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/8271/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-8271/

File: templates/federated_reporting/distributed_cleanup.py [hunk 1]
```diff
@@ -120,11 +128,16 @@ def interactive_setup_feeder(hub, email, fr_distributed_cleanup_password):
         sys.exit(1)
 
 
-def interactive_setup():
-    fr_distributed_cleanup_password = "".join(random.choices(string.printable, k=20))
-    admin_pass = getpass(
-        prompt="Enter admin password for superhub {}: ".format(socket.getfqdn())
-    )
+def interactive_setup(force_interactive=False):
+    fr_distributed_cleanup_password = "".join(random.choices(string.digits + string.punctuation + string.ascii_letters, k=20))
```
Comment (Lex-2008) [inline templates/federated_reporting/distributed_cleanup.py:132]: I wonder if other symbols in `string.punctuation` can cause similar issues in API usage? If it goes through HTTP request, they're supposed to be %-encoded-decoded and that's where the space issue might be coming from. Maybe it's safer to use `string.digits + string.ascii_letters`? I believe we can increase `k=20` to something bigger for extra security :)

**Review by Lex-2008 — APPROVED** [review]

**Comment by craigcomstock** [comment]
packages are ready for 8271 so requested a fr-tests with my changes there [Build Status](https://ci.cfengine.com/job/fr-tests/4665/)

**Comment by craigcomstock** [comment]
build wasn't getting my context change in the python script from "cfengine" to "cfengine_3" so fr-tests was still failing.

Retrying build with "pull/2424/head" instead of just "2424" in parameters: [Build Status](https://ci.cfengine.com/job/pr-pipeline/8274/)

**Comment by craigcomstock** [comment]
Ok! Local fr-test worked. Trying full CI run of fr-tests with latest packages: [Build Status](https://ci.cfengine.com/job/fr-tests/4669/)

File: templates/federated_reporting/distributed_cleanup.py (hunk 1 above)
Comment (Lex-2008) [reply templates/federated_reporting/distributed_cleanup.py:132]: ok! Nah, no need to investigate too much, I was just wondering if any of the "funny characters" in `string.punctuation` can cause us an issue. But finding why exactly space causes the breakage equals to fixing it :)

I'm just thinking: maybe password gets pasted into some shell command, like
```
sh -c "cleanup.sh $USER $PASSWORD"
```
then not only space, but a pipe character will also cause breakage!

**Comment by craigcomstock** [comment]
bug fix for ENT-8979, multiple deletes on a feeder results in malformed SQL, so @cf-bottom gimme another build and I'll re-run fr-tests on it tomorrow morning.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/8278/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/8278/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-8278/

File: templates/federated_reporting/distributed_cleanup.py (hunk 1 above)
Comment (craigcomstock) [reply templates/federated_reporting/distributed_cleanup.py]: I just removed punctuation, should be OK without it.

File: cfe_internal/enterprise/federation/federation.cf [hunk 2]
```diff
@@ -885,10 +885,13 @@ bundle agent distributed_cleanup_setup
   files:
     am_superhub|am_feeder::
       "${distributed_cleanup_dir}/."
+        perms => default:mog( "700", "$(cfengine_enterprise_federation:config.transport_user)", "root" ),
```
Comment (nickanderson) [inline cfe_internal/enterprise/federation/federation.cf:888]: +1

I think this one should be backported.

File: templates/federated_reporting/distributed_cleanup.py [hunk 3]
```diff
@@ -120,11 +128,16 @@ def interactive_setup_feeder(hub, email, fr_distributed_cleanup_password):
         sys.exit(1)
 
 
-def interactive_setup():
-    fr_distributed_cleanup_password = "".join(random.choices(string.printable, k=20))
-    admin_pass = getpass(
-        prompt="Enter admin password for superhub {}: ".format(socket.getfqdn())
-    )
+def interactive_setup(force_interactive=False):
+    fr_distributed_cleanup_password = "".join(random.choices(string.digits + string.ascii_letters, k=20))
```
Comment (nickanderson) [inline templates/federated_reporting/distributed_cleanup.py:132]: What is this fr_distributed_cleanup_password bit with random choices?

File: templates/federated_reporting/distributed_cleanup.py [hunk 4]
```diff
@@ -385,11 +385,13 @@ def main():
         # simulate the host api delete process by setting current_timestamp in deleted column
         # and delete from all federated tables similar to the clear_hosts_references() pgplsql function.
         post_sql += "INSERT INTO __hosts (hostkey,deleted) VALUES"
+        deletes = []
         for hostkey in post_hostkeys:
-            delete_sql += "('{}', CURRENT_TIMESTAMP) ".format(hostkey)
+            deletes.append("('{}', CURRENT_TIMESTAMP)".format(hostkey))
 
+        delete_sql = ", ".join(deletes)
```
Comment (nickanderson) [inline templates/federated_reporting/distributed_cleanup.py:392]: ack

**Review by nickanderson — APPROVED** [review]
I don't know how valuable my review is with respect to the python but other than one question seems ok to me.

File: templates/federated_reporting/distributed_cleanup.py (hunk 3 above)
Comment (craigcomstock) [reply templates/federated_reporting/distributed_cleanup.py:132]: This is the password for the account created and used later by this utility. The password is saved via cf-secret, encrypted to only the superhub.

File: cfe_internal/enterprise/federation/federation.cf (hunk 2 above)
Comment (craigcomstock) [reply cfe_internal/enterprise/federation/federation.cf:888]: I was planning on back-porting all of this to 3.18.x where distributed cleanup is present. Not 3.15, didn't get it back that far.

**Review by vpodzime — APPROVED** [review]
Looks good to me. It looks like the last commit actually makes things work properly -- should we make it available as a patch for the recently-released versions?

**Review by Lex-2008 — APPROVED** [review]

**Comment by craigcomstock** [comment]
> Looks good to me. It looks like the last commit actually makes things work properly -- should we make it available as a patch for the recently-released versions?

Yes. I logged a ticket as a blocker and intend on informing/sending a patch to customers who we know are using or have used or trying this facility. :+1:

**Comment by craigcomstock** [comment]
cherry picked to 3.18.x https://github.com/cfengine/masterfiles/pull/2427

**Pull request merged**